### PR TITLE
Firebase integration issue

### PR DESCRIPTION
**Pull request by ntunin**
## Description
There I placed all available broads in ButterBroad with their tests and integration instructions. other repositories can be removed. Also I've added `AnyCodable` as source code to decrease integration complexity

## Related tasks
[ISD-326](https://rosberry.atlassian.net/browse/ISD-326)

## Checklist

- [X] Have merged develop branch
- [X] Develop branch is chosen as a merge branch
- [X] The appropriate pull request label is selected (Ready | WIP)
- [X] Follows code style
- [ ] Has no SwiftLint warnings (There is warnings from 3rd party source code)
- [X] Retain cycles were checked
- [X] Has no force unwraps
- [X] Access control words were checked
- [X] All tests passed succesfully
- [X] All example projects runs and hadle all events correctly
- [X] Checked that the code correctly handles `background` and `foreground` states

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,33 +27,170 @@ Drag `Sources` folder from [last release](https://github.com/rosberry/butterbroa
...
+### Firebase broad
+
+To integrate `FirebaseBroad` into your project you need to integrate `Firebase` and `FirebaseAnalytics` by any awailable way before. For example there is an [oficial interation instuction](https://firebase.google.com/docs/analytics/get-started?platform=ios) from Firebase. If you preffer `Carthage` then you can use folowing binaries:
+
+```
+binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseAnalyticsBinary.json"
+binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseCrashlyticsBinary.json"
+```
+
+Unlike `Pod installation` `FirebaseCrashlytics.framework`  builded via carthage has not required  `run` and `upload-symbols` scripts, so we packed them into `ButterBroad.framework` for your convenience. To use them create run script phase `Crashlytics` with content
```
Comment (artemnovichkov) [inline README.md:85]:
```suggestion
Unlike `Pod installation` `FirebaseCrashlytics.framework`  built via carthage has not required  `run` and `upload-symbols` scripts, so we packed them into `ButterBroad.framework` for your convenience. To use them create run script phase `Crashlytics` with content
```

File: README.md [hunk 2]
```diff
@@ -27,33 +27,170 @@ Drag `Sources` folder from [last release](https://github.com/rosberry/butterbroa
...
+
+import ButterBroad
+
+extension Butter {
+    static let firebase: FirebaseBroad = .init(with: .default)
+    static let common: Butter = .init(broads: analog)
+}
+```
+
+You can put one of the folowing activation types into initializer:
```
Comment (artemnovichkov) [inline README.md:110]:
```suggestion
You can put one of the following activation types into initializer:
```

File: README.md [hunk 3]
```diff
@@ -27,33 +27,170 @@ Drag `Sources` folder from [last release](https://github.com/rosberry/butterbroa
...
+
+- `.none`: No action requered.  `FirebaseApp.configure`  should be triggerd from the application. Use it if your app has other firebase features and firbase app configuration is your strict responsibility. This value is asumed by default.
```
Comment (artemnovichkov) [inline README.md:112]:
```suggestion
- `.none`: No action required.  `FirebaseApp.configure`  should be triggered from the application. Use it if your app has other firebase features and firebase app configuration is your strict responsibility. This value is assumed by default.
```

File: README.md [hunk 4]
```diff
@@ -27,33 +27,170 @@ Drag `Sources` folder from [last release](https://github.com/rosberry/butterbroa
...
+- `.default`: perform `FirebaseApp.configure()` in the `FirebaseBroad`. Use it if you do not use other firebase features.
+- `.custom`: perform specific `FirebaseApp.configure` in the `FirebaseBroad`. Use it if you do not use other firebase features but you need to specify some configuration parameters. 
+
+### Facebook broad
+
+To integrate `FacebookBroad` into your project you need to integrate `Facebook` and `FacebookAnalytics` by [any awailable way](https://developers.facebook.com/docs/analytics/quickstart-list/ios/) before.
+
+Then copy [FacebookBroad](https://github.com/rosberry/butterbroad/tree/Sources/Broads/FacebookBroad.swift) into your project and put an `FacebookBroad` instance into  `ButterBroad`:
```
Comment (artemnovichkov) [inline README.md:120]:
```suggestion
Then copy [FacebookBroad](https://github.com/rosberry/butterbroad/tree/Sources/Broads/FacebookBroad.swift) into your project and put an `FacebookBroad` instance into `ButterBroad`:
```

File: README.md [hunk 5]
```diff
@@ -27,33 +27,170 @@ Drag `Sources` folder from [last release](https://github.com/rosberry/butterbroa
...
+    static let facebook: FacebookBroad = .init(with: .default)
+    static let common: Butter = .init(broads: facebook)
+}
+```
+
+You can put one of the folowing activation types into initializer:
+
+- `.none`: No action requered.  `AppEvents.activateApp`  should be triggerd from the application. Use it if your app has other facebook features and facebook app configuration is your strict responsibility. This value is asumed by default.
+- `.default`: perform `AppEvents.activateApp()` in the `FacebookBroad`. Use it if you do not use other facebook features.
+- `.custom`: perform specific `AppEvents.activateApp` in the `FacebookBroad`. Use it if you do not use other facebook features but you need to specify some configuration parameters.
```
Comment (artemnovichkov) [inline README.md:137]:
```suggestion
You can put one of the following activation types into initializer:

- `.none`: No action required.  `AppEvents.activateApp`  should be triggered from the application. Use it if your app has other facebook features and facebook app configuration is your strict responsibility. This value is asumed by default.
- `.default`: perform `AppEvents.activateApp()` in the `FacebookBroad`. Use it if you do not use other facebook features.
- `.custom`: perform specific `AppEvents.activateApp` in the `FacebookBroad`. Use it if you do not use other facebook features but you need to specify some configuration parameters. 
```

File: README.md [hunk 6]
```diff
@@ -27,33 +27,170 @@ Drag `Sources` folder from [last release](https://github.com/rosberry/butterbroa
...
+class AppDelegate: UIResponder, UIApplicationDelegate {
+    func applicationDidBecomeActive(_ application: UIApplication) {
+        Butter.facebook.activate()
+    }
+}
+```
+
+### User defined broad
+
+To specify your own analytics agreggator create class that implements `ButterBroad.Analytics` protocol and specify `public func log(_ event: Event)` method to send data to some analytics. `Event` contains `String` name and `[String: AnyCodable]` parameters, where [`AnyCodable`](https://github.com/Flight-School/AnyCodable/tree/master/Sources/AnyCodable) is type erraser for `Codable` generic protocol. If the analytics assumes some activation you can specify `public func activate()` method. To do it in more convenient way `ButterBroad` contains `enum Activation`. Look on the `FirebaseBroad` implementation for example
```
Comment (artemnovichkov) [inline README.md:152]:
```suggestion
To specify your own analytics aggregator create class that implements `ButterBroad.Analytics` protocol and specify `public func log(_ event: Event)` method to send data to some analytics. `Event` contains `String` name and `[String: AnyCodable]` parameters, where [`AnyCodable`](https://github.com/Flight-School/AnyCodable/tree/master/Sources/AnyCodable) is type eraser for `Codable` generic protocol. If the analytics assumes some activation you can specify `public func activate()` method. To do it in a more convenient way `ButterBroad` contains `enum Activation`. Look on the `FirebaseBroad` implementation for example
```

File: Sources/Broads/FirebaseBroad.swift [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+final public class FirebaseBroad: ButterBroad.Analytics {
+
+    private let activation: Activation
+
+    /// Creates `FirebaseBroad` instance with provided activation type
+    ///
+    /// - parameter activation: specifies an action for the Facebook that should be aplied on `FirebaseBroad`activation. The default value `.none` assumes that
+    /// no action required and `FirebaseApp.configure` will called on the user side. The value `.default` assumes that `FirebaseApp.configure()` will called
+    /// on the `FirebaseBroad` side. The value `.custom` assumes that user provided specific activation will called on the `FirebaseBroad` side.
+    public init(with activation: Activation = .none) {
```
Comment (artemnovichkov) [inline Sources/Broads/FirebaseBroad.swift:19]:
```suggestion
    public init(activation: Activation = .none) {
```

File: Sources/Broads/FacebookBroad.swift [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+public final class FacebookBroad: Analytics {
+
+    private let activation: Activation
+
+    /// Creates `FacebookBroad` instance with provided activation type
+    ///
+    /// - parameter activation: specifies an action for the Facebook that should be aplied on `ButterBroad`activation. The default value `.none` assumes that
+    /// no action required and `AppEvents.activateApp` will called on the user side. The value `.default` assumes that `AppEvents.activateApp()` will called
+    /// on the `FacebookBroad` side. The value `.custom` assumes that user provided specific activation will called on the `FacebookBroad` side.
+    public init(with activation: Activation = .none) {
```
Comment (artemnovichkov) [inline Sources/Broads/FacebookBroad.swift:20]:
```suggestion
    public init(activation: Activation = .none) {
```

**Comment by ntunin** [comment]
@Antowkos, @stason314

File: ButterBroadExample/Cartfile [hunk 9]
```diff
@@ -1,4 +1,5 @@
-github "rosberry/ButterBroad"
-github "rosberry/AnalogBroad"
-github "rosberry/FaceBookBroad"
-github "rosberry/CrashlyticsBroad"
+github "https://github.com/rosberry/ButterBroad" "hotfix/ISD-326"
```
Comment (disabled) [inline ButterBroadExample/Cartfile:1]: Don't forget to switch away from hotfix branch after merge.

**Pull request merged**